index.bs

```diff
@@ -1747,28 +2838,94 @@ partial interface MLGraphBuilder {
-            - *activations*: a sequence of {{MLActivation}}. A pair of activation functions with the first function used for the update and reset gate, and the second used for the new gate. When not specified, it's assumed to be the sigmoid (*"sigmoid"*) and the hyperbolic tangent (*"tanh"*) function respectively.
 
-    **Returns:** a sequence of {{MLOperand}}. The first element of the sequence is a 3-D tensor of shape [num_directions, batch_size, hidden_size], the cell output from the last time step of the network. Additionally, if *options.returnSequence* is set to true, the second element is the 4-D output tensor of shape [steps, num_directions, batch_size, hidden_size] containing every cell outputs from each time step in the temporal sequence.
+    **Returns:** a sequence of {{MLOperand}}. The first element of the sequence is a 3-D tensor of shape [numDirections, batchSize, hiddenSize], the cell output from the last time step of the network. Additionally, if |options|.{{MLGruOptions/returnSequence}} is set to `true`, the second element is the 4-D output tensor of shape [steps, numDirections, batchSize, hiddenSize] containing every cell outputs from each time step in the temporal sequence.
+</div>
+
+<details open algorithm>
+
+  <summary>
+    The <dfn method for=MLGraphBuilder>gru(|input|, |weight|, |recurrentWeight|, |steps|, |hiddenSize|, |options|)</dfn> method steps are:
+  </summary>
+  <div class=algorithm-steps>
+    1. [=Assert=]: the type of |input|, |weight| and |recurrentWeight| is {{MLOperand}}.
+    1. If the [=rank=] of |input| or |weight| is not `3`, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. If the [=rank=] of |weight| or |recurrentWeight| is not `2`, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. If |options|.{{MLGruOptions/bias}} [=map/exists=].
+        1. [=Assert=]: its type is {{MLOperand}}.
+        1. If its rank is not `2`, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. If |options|.{{MLGruOptions/recurrentBias}} [=map/exists=].
+        1. [=Assert=]: its type is {{MLOperand}}.
+        1. If its rank is not `2`, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. If |options|.{{MLGruOptions/initialHiddenState}} [=map/exists=].
+        1. [=Assert=]: its type is {{MLOperand}}.
+        1. If its rank is not `3`, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. If |options|.{{MLGruOptions/direction}} is not one of {{MLRecurrentNetworkDirection}}, then [=exception/throw=] a {{TypeError}}.
+    1. If |options|.{{MLGruOptions/layout}} is not one of {{MLGruWeightLayout}}, then [=exception/throw=] a {{TypeError}}.
+    1. If |options|.{{MLGruOptions/activations}} [=map/exists=] and is not an array of {{MLActivation}} objects with size `2`, then [=exception/throw=] a {{TypeError}}.
```

Comment:
Type checks taken care of by WebIDL not necessary, so this can be `... [=map/exists=] and its [=list/size=] is not 2...`

(Aside: the 2 would not normally be styled as code, it's part of the algorithm and no longer a JS value)